File: src/graphs/dfs/lc_841_keys_and_rooms/keys_and_rooms.py
class Solution:
    def canVisitAllRooms(self, rooms: list[list[int]]) -> bool:
        adj = {}
        def dfs(node, visited):
            if node in visited:
                return
            visited.add(node)
            if node in adj:
                for neigh in adj[node]:
                    if neigh not in visited:
                        dfs(neigh, visited)
        for i, room in enumerate(rooms):
            adj[i] = room
        visited = set()
        
        dfs(0, visited)
        return len(visited) == len(rooms)

File: src/graphs/dfs/lc_841_keys_and_rooms/keys_and_rooms.go
func canVisitAllRooms(rooms [][]int) bool {
    adj := make(map[int][]int)
    visited := make(map[int]bool)
    for i, room := range rooms {
        adj[i] = room
    }
    dfs(adj, 0, visited)
    return len(visited) == len(rooms)
}

func dfs(adj map[int][]int, node int, visited map[int]bool) {
    _, ispresent := visited[node]
    if ispresent {
        return
    }
    visited[node] = true
    _, isNode := adj[node]
    if isNode {
        rooms := adj[node]
        for _, neigh := range(rooms) {
            dfs(adj, neigh, visited)
        }
    }
}